### DISCUSSION TITLE
Invalid Auth Error with Home Assistant

### DISCUSSION THREAD

**Discussion by lundyfpv** [toverainc/willow#97] [Q&A] [answered]
I'm sure the issue is on my end but I have tried tons of different methods to get this to work but they all fail with a 401 error in the serial logs. Any idea what I could be doing wrong. I do have my HA instance behind a Cloudflare tunnel but curl seems to work fine for calling services. 

**Tio Serial Stream:**
`I (01:29:23.910) WILLOW: AUDIO_REC_VAD_START
E (01:29:23.911) AUDIO_ELEMENT: [http_stream_writer] Element already stopped
E (01:29:23.912) AUDIO_ELEMENT: [raw_stream_writer_to_api] Element already stopped
I (01:29:23.921) AUDIO_THREAD: The http_stream_writer task allocate stack on external memory
I (01:29:23.931) AUDIO_ELEMENT: [http_stream_writer-0x3d8146a0] Element task created
I (01:29:23.938) AUDIO_ELEMENT: [raw_stream_writer_to_api-0x3d8144b0] Element task created

I (01:29:23.947) AUDIO_PIPELINE: Func:audio_pipeline_run, Line:359, MEM Total:7128631 Bytes, Inter:135931 Bytes, Dram:135931 Bytes
I (01:29:23.960) AUDIO_ELEMENT: [http_stream_writer] AEL_MSG_CMD_RESUME,state:1
I (01:29:23.968) AUDIO_PIPELINE: Pipeline started
audio_pipeline_run(hdl_ap_to_api) - uri: 'https://infer.tovera.io/api/willow'
I (01:29:23.982) WILLOW: [ + ] HTTP client HTTP_STREAM_PRE_REQUEST, length=0
I (01:29:24.269) WILLOW: AUDIO_REC_VAD_END
I (01:29:24.270) WILLOW: AUDIO_REC_WAKEUP_END
Total bytes written: 4096
Total bytes written: 8192
Total bytes written: 12288
Total bytes written: 16384
Total bytes written: 20480
Total bytes written: 22528
I (01:29:24.674) AUDIO_ELEMENT: IN-[http_stream_writer] AEL_IO_DONE,-2
I (01:29:24.675) WILLOW: [ + ] HTTP client HTTP_STREAM_POST_REQUEST, write end chunked marker
I (01:29:25.059) HTTP_CLIENT: Body received in fetch header state, 0x3d953d63, 31
I (01:29:25.060) WILLOW: [ + ] HTTP client HTTP_STREAM_FINISH_REQUEST
I (01:29:25.064) WILLOW: Got HTTP Response = {"text":"Bye!","language":"en"}
I (01:29:25.072) WILLOW: HASS URL: https://ha.mysite.com:8443/api/conversation/process
I (01:29:25.080) WILLOW: sending '{"text":"Bye!","language":"en"}' to Home Assistant API on 'https://ha.mysite.com:8443/api/conversation/process'
I (01:29:25.919) HTTP_CLIENT: Body received in fetch header state, 0x3d952570, 17
I (01:29:25.920) WILLOW: HTTP status='401' content_length='17'
I (01:29:25.927) AUDIO_THREAD: The play_tone_err task allocate stack on external memory
I (01:29:25.933) WILLOW: HTTP POST response body:
401
I (01:29:27.942) AUDIO_ELEMENT: [http_stream_writer] AEL_MSG_CMD_PAUSE
I (01:29:35.939) WILLOW: Wake LCD timeout, turning off LCD`

**HA Logs:**
`Login attempt or request with invalid authentication from IPaddress (IPaddress). Requested URL: '/api/conversation/process'. (ESP32 HTTP Client/1.0)`

And because I was starting to go crazy I also tried using curl with my Long Lived Access Token and that appears to work fine and turn on my input boolean.

`curl -X POST -H "Authorization: Bearer 123456......." -H "Content-Type: application/json" -d '{"entity_id":"input_boolean.test"}' "https://ha.mysite.com:8443/api/services/input_boolean/turn_on"`

**Comment by kristiankielhofner** [comment]
Can you provide your curl command with verbose output with the creds removed?

Also, I assume the machine you're running curl on has the WARP client connected to the tunnel? The ESP BOX doesn't have that so for an equivalent test you will need to make sure you have WARP disconnected on the machine you're running curl from.

**Comment by lundyfpv** [comment]
Thanks for the fast response. I am not running WARP client just using the Cloudflare tunnel for protecting my HA instance. 

**Verbose Curl Output:**

`Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 104.21.77.241:8443...
* Connected to ha.mysite.com (ipAddress) port 8443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: /etc/ssl/certs
* TLSv1.0 (OUT), TLS header, Certificate Status (22):
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS header, Certificate Status (22):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS header, Finished (20):
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.2 (OUT), TLS header, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN, server accepted to use h2
* Server certificate:
*  subject: C=US; ST=California; L=San Francisco; O=Cloudflare, Inc.; CN=sni.cloudflaressl.com
*  start date: Jan  1 00:00:00 2023 GMT
*  expire date: Jan  1 23:59:59 2024 GMT
*  subjectAltName: host "ha.mysite.com" matched cert's "*.mysite.com"
*  issuer: C=US; O=Cloudflare, Inc.; CN=Cloudflare Inc ECC CA-3
*  SSL certificate verify ok.
* Using HTTP2, server supports multiplexing
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* Using Stream ID: 1 (easy handle 0xaaab1c3ab040)
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
> POST /api/services/input_boolean/turn_on HTTP/2
> Host: ha.mysite.com:8443
> user-agent: curl/7.81.0
> accept: */*
> authorization: Bearer 111111111111111111111111111111111111111111.......
> content-type: application/json
> content-length: 34
>
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* We are completely uploaded and fine
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* old SSL session ID is stale, removing
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* Connection state changed (MAX_CONCURRENT_STREAMS == 256)!
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* TLSv1.2 (IN), TLS header, Supplemental data (23):
< HTTP/2 403
< date: Tue, 23 May 2023 19:57:35 GMT
< content-type: text/plain; charset=utf-8
< content-length: 14
< cf-cache-status: DYNAMIC
< report-to: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v3?s=Q72lABKWyGRE3eOH%2Fyo%2B%2FZi7rHmKRGX7kYRwg6K6L1nnZMsjcJtJT2QzVyD7cU6C5fUUu9pXlZxzyZdeAQhbeO1uZ9RXXpYEJ%2F3Ipi8iQiMYbYAZ%2BTDMOX8%2F0FFYmblC%2F%2FUaOHsqfg%3D%3D"}],"group":"cf-nel","max_age":604800}
< nel: {"success_fraction":0,"report_to":"cf-nel","max_age":604800}
< server: cloudflare
< cf-ray: 7cbfde6a2c993bd9-MEM
< alt-svc: h3=":8443"; ma=86400, h3-29=":8443"; ma=86400
<
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* Connection #0 to host ha.mysite.com left intact`

**Comment by kristiankielhofner** [comment]
Your example curl is returning 403. Why is that?

I'm also struggling to understand this config (and purpose)... Is your HA instance not local? We've never tested the Willow http client against Cloudflare because frankly we never considered this use case. Theoretically it should "just work" but I have a lot of experience with CF and CF tunnels and I can think of any number of settings/config options that could potentially be problematic.

Can you post a de-authenticated successful curl with ```-v --http1.1 --tlsv1.2``` ?

**Comment by lundyfpv** [comment]
So might have been a few things. My ip was banned in HA from too many failed auth requests and also apparently I need to add -v at the end of my curl command. So the reason for the Coudflare tunnel is for secure-ish remote access but doing this also prevents local http requests from what I can tell so I have to use my URL that matches my cert. Thanks again for the help though!

Here is an explanation of the Cloudflare tunnel: https://everythingsmarthome.co.uk/free-remote-access-for-home-assistant-with-cloudlfare/

**When running:** curl -X POST -H "Authorization: Bearer 111111111111111" -H "Content-Type: application/json" -d '{"entity_id":"input_boolean.test"}' "https://ha.mysite.com:8443/api/services/input_boolean/turn_on" -v --http1.1 --tlsv1.2

**I get:** 
*   Trying ipAddress:8443...
* Connected to ha.mysite.com (ipAddress) port 8443 (#0)
* ALPN, offering http/1.1
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: /etc/ssl/certs
* TLSv1.0 (OUT), TLS header, Certificate Status (22):
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS header, Certificate Status (22):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS header, Finished (20):
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.2 (OUT), TLS header, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN, server accepted to use http/1.1
* Server certificate:
*  subject: C=US; ST=California; L=San Francisco; O=Cloudflare, Inc.; CN=sni.cloudflaressl.com
*  start date: Jan  1 00:00:00 2023 GMT
*  expire date: Jan  1 23:59:59 2024 GMT
*  subjectAltName: host "ha.mysite.com" matched cert's "*.mysite.com"
*  issuer: C=US; O=Cloudflare, Inc.; CN=Cloudflare Inc ECC CA-3
*  SSL certificate verify ok.
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
> POST /api/services/input_boolean/turn_on HTTP/1.1
> Host: ha.mysite.com:8443
> User-Agent: curl/7.81.0
> Accept: */*
> Authorization: Bearer 111111111111111111111111111111111111
> Content-Type: application/json
> Content-Length: 34
>
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* old SSL session ID is stale, removing
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Date: Tue, 23 May 2023 20:50:26 GMT
< Content-Type: application/json
< Transfer-Encoding: chunked
< Connection: keep-alive
< CF-Cache-Status: DYNAMIC
< Report-To: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v3?s=uHNPW%2Fg80IO1bYvZopY6RDlAKfEHditN5avyUmQicfcO82QfbadgvADxudPfhgR5eOalE1O%2F2RXHm%2Bm4YX6YM2WS9zo3BPdfWBcoczfIBygBgx%2Fsthq9CeNjH5%2FGHgI3czliAAPhag%3D%3D"}],"group":"cf-nel","max_age":604800}
< NEL: {"success_fraction":0,"report_to":"cf-nel","max_age":604800}
< Server: cloudflare
< CF-RAY: 7cc02bd4285d0deb-MEM
< alt-svc: h3=":8443"; ma=86400, h3-29=":8443"; ma=86400
<
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* Connection #0 to host ha.mysite.com left intact
[{"entity_id":"input_boolean.test","state":"on","attributes":{"editable":true,"friendly_name":"test"},"last_changed":"2023-05-23T20:50:26.717845+00:00","last_updated":"2023-05-23T20:50:26.717845+00:00","context":{"id":"01H156188WMRF52760M5Z53T2E","parent_id":null,"user_id":"ad088ed07af74937a14c74482e36811f"}}]

**Reply by kristiankielhofner** [reply]
So to confirm - everything is resolved and working now?

**Reply by lundyfpv** [reply]
No sadly still getting invalid authentication from the Willow side. My curl request with the exact same Access Token works fine as above. That's what's got me pulling my hair out. 

HA still says 'Login attempt or request with invalid authentication from ipAddress. Requested URL: '/api/conversation/process'. (ESP32 HTTP Client/1.0)'

Willow Serial Log still says:
'I (04:32:10.348) WILLOW: AUDIO_REC_WAKEUP_START

I (04:32:11.451) WILLOW: AUDIO_REC_VAD_START
E (04:32:11.452) AUDIO_ELEMENT: [http_stream_writer] Element already stopped
E (04:32:11.453) AUDIO_ELEMENT: [raw_stream_writer_to_api] Element already stopped
I (04:32:11.462) AUDIO_THREAD: The http_stream_writer task allocate stack on external memory
I (04:32:11.471) AUDIO_ELEMENT: [http_stream_writer-0x3d8156c0] Element task created
I (04:32:11.479) AUDIO_ELEMENT: [raw_stream_writer_to_api-0x3d81461c] Element task created
I (04:32:11.488) AUDIO_PIPELINE: Func:audio_pipeline_run, Line:359, MEM Total:7129023 Bytes, Inter:135931 Bytes, Dram:135931 Bytes

I (04:32:11.501) AUDIO_ELEMENT: [http_stream_writer] AEL_MSG_CMD_RESUME,state:1
I (04:32:11.509) AUDIO_PIPELINE: Pipeline started
audio_pipeline_run(hdl_ap_to_api) - uri: 'https://infer.tovera.io/api/willow'
I (04:32:11.522) WILLOW: [ + ] HTTP client HTTP_STREAM_PRE_REQUEST, length=0
Total bytes written: 4096
Total bytes written: 8192
Total bytes written: 12288
Total bytes written: 16384
Total bytes written: 20480
Total bytes written: 24576
Total bytes written: 28672
Total bytes written: 32768
Total bytes written: 36864
Total bytes written: 40960
Total bytes written: 45056
I (04:32:12.655) WILLOW: [ + ] HTTP client HTTP_STREAM_ON_REQUEST, length=4096
I (04:32:12.680) WILLOW: AUDIO_REC_VAD_END
Total bytes written: 49152
Total bytes written: 51200
I (04:32:12.731) AUDIO_ELEMENT: IN-[http_stream_writer] AEL_IO_DONE,-2
I (04:32:12.732) WILLOW: [ + ] HTTP client HTTP_STREAM_POST_REQUEST, write end chunked marker
I (04:32:13.104) HTTP_CLIENT: Body received in fetch header state, 0x3d954123, 50
I (04:32:13.105) WILLOW: [ + ] HTTP client HTTP_STREAM_FINISH_REQUEST
I (04:32:13.109) WILLOW: Got HTTP Response = {"text":"Turn on kitchen lights.","language":"en"}
I (04:32:13.119) WILLOW: HASS URL: https://ha.mysite.com:8443/api/conversation/process
I (04:32:13.127) WILLOW: sending '{"text":"Turn on kitchen lights.","language":"en"}' to Home Assistant API on 'https://ha.mysite.com:8443/api/conversation/process'
I (04:32:14.171) HTTP_CLIENT: Body received in fetch header state, 0x3d95972c, 17
I (04:32:14.172) WILLOW: HTTP status='401' content_length='17'
I (04:32:14.179) AUDIO_THREAD: The play_tone_err task allocate stack on external memory
I (04:32:14.185) WILLOW: HTTP POST response body:
401
I (04:32:16.195) AUDIO_ELEMENT: [http_stream_writer] AEL_MSG_CMD_PAUSE
I (04:32:24.191) WILLOW: Wake LCD timeout, turning off LCD'

**Reply by lundyfpv** [reply]
For what it's worth I modified my curl command to send to the conversation api and it did work as expected.

curl -X POST -H "Authorization: Bearer 1111111111111111111111111111111111" -H "Content-Type: application/json" -d '{"text":"Turn on kitchen lights.","language":"en"}' "https://ha.mysite.com:8443/api/services/conversation/process" -v
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying ipAddress:8443...
* Connected to ha.mysite.com (ipAddress) port 8443 (#0)
* ALPN: offers h2
* ALPN: offers http/1.1
*  CAfile: /usr/local/share/certs/ca-root-nss.crt
*  CApath: none
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN: server accepted h2
* Server certificate:
*  subject: C=US; ST=California; L=San Francisco; O=Cloudflare, Inc.; CN=sni.cloudflaressl.com
*  start date: Jan  1 00:00:00 2023 GMT
*  expire date: Jan  1 23:59:59 2024 GMT
*  subjectAltName: host "ha.mysite.com" matched cert's "*.mysite.com"
*  issuer: C=US; O=Cloudflare, Inc.; CN=Cloudflare Inc ECC CA-3
*  SSL certificate verify ok.
* Using HTTP2, server supports multiplexing
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* h2h3 [:method: POST]
* h2h3 [:path: /api/services/conversation/process]
* h2h3 [:scheme: https]
* h2h3 [:authority: ha.mysite.com:8443]
* h2h3 [user-agent: curl/7.85.0]
* h2h3 [accept: */*]
* h2h3 [authorization: Bearer 111111111111111111111111111
* h2h3 [content-type: application/json]
* h2h3 [content-length: 50]
* Using Stream ID: 1 (easy handle 0x8012d6800)
> POST /api/services/conversation/process HTTP/2
> Host: ha.mysite.com:8443
> user-agent: curl/7.85.0
> accept: */*
> authorization: Bearer 1111111111111111111111
> content-type: application/json
> content-length: 50
>
* We are completely uploaded and fine
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* old SSL session ID is stale, removing
* Connection state changed (MAX_CONCURRENT_STREAMS == 256)!
< HTTP/2 200
< date: Tue, 23 May 2023 22:58:21 GMT
< content-type: application/json
< cf-cache-status: DYNAMIC
< report-to: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v3?s=9etsraOQBzZ7%2BHl343cy%2BoCNgZLukwrx9RFz0TafqfvImZnMKPjclNUHsCFR3t8Ueq4MOz6kewF7%2FG11QS3peInLPPu4gKAThpl2PNUC2meADwQYJclEQX4qYqWnzA0s3Hxb5KdqNQ%3D%3D"}],"group":"cf-nel","max_age":604800}
< nel: {"success_fraction":0,"report_to":"cf-nel","max_age":604800}
< server: cloudflare
< cf-ray: 7cc0e730eea429a5-ORD
< alt-svc: h3=":8443"; ma=86400, h3-29=":8443"; ma=86400
<
* Connection #0 to host ha.mysite.com left intact

**Reply by kristiankielhofner** [reply]
Please try the curl command I provided earlier with ```-v --http1.1 --tlsv1.2``` and provide the output.

**Reply by lundyfpv** [reply]
Sure thing thought I had already posted it. Sorry I'm not super knowledgeable about all of this I just know enough to be dangerous and screw up stuff so I hope this is what you are looking for. 

**Turning on an input boolean:**

curl -X POST -H "Authorization: Bearer 1111111111111" -H "Content-Type: application/json" -d '{"entity_id":"input_boolean.test"}' "https://ha.mysite.com:8443/api/services/input_boolean/turn_on" -v --http1.1 --tlsv1.2
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying ipAddress:8443...
* Connected to ha.mysite.com (ipAddress) port 8443 (#0)
* ALPN: offers http/1.1
*  CAfile: /usr/local/share/certs/ca-root-nss.crt
*  CApath: none
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN: server accepted http/1.1
* Server certificate:
*  subject: C=US; ST=California; L=San Francisco; O=Cloudflare, Inc.; CN=sni.cloudflaressl.com
*  start date: Jan  1 00:00:00 2023 GMT
*  expire date: Jan  1 23:59:59 2024 GMT
*  subjectAltName: host "ha.mysite.com" matched cert's "*.mysite.com"
*  issuer: C=US; O=Cloudflare, Inc.; CN=Cloudflare Inc ECC CA-3
*  SSL certificate verify ok.
> POST /api/services/input_boolean/turn_on HTTP/1.1
> Host: ha.mysite.com:8443
> User-Agent: curl/7.85.0
> Accept: */*
> Authorization: Bearer 11111111111111
> Content-Type: application/json
> Content-Length: 34
>
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* old SSL session ID is stale, removing
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Date: Wed, 24 May 2023 01:36:33 GMT
< Content-Type: application/json
< Transfer-Encoding: chunked
< Connection: keep-alive
< CF-Cache-Status: DYNAMIC
< Report-To: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v3?s=%2B2KztgxduXHUA1figq8ueW4DFOCntwutfn%2BLNG0JBreCgUdkaKIpeCxIDhHkyMX2IcApu3kxRDxUDcTa8yUN2HQyAiBwrePF4XGKL8KPMTfaOGG33%2FKg7NZv6ffmJJ%2BFMwAA1pLRjw%3D%3D"}],"group":"cf-nel","max_age":604800}
< NEL: {"success_fraction":0,"report_to":"cf-nel","max_age":604800}
< Server: cloudflare
< CF-RAY: 7cc1ceee0c118140-ORD
< alt-svc: h3=":8443"; ma=86400, h3-29=":8443"; ma=86400
<
* Connection #0 to host ha.mysite.com left intact
[{"entity_id":"input_boolean.test","state":"on","attributes":{"editable":true,"friendly_name":"test"},"last_changed":"2023-05-24T01:36:33.088814+00:00","last_updated":"2023-05-24T01:36:33.088814+00:00","context":{"id":"01H15PD49R1D0Z3C7PDDX4T608","parent_id":null,"user_id":"ad088ed07af74937a14c74482e36811f"}}]%


**Turning on lights with the conversation API:**

curl -X POST -H "Authorization: Bearer 11111111111111111111111" -H "Content-Type: application/json" -d '{"text":"Turn on kitchen lights.","language":"en"}' "https://ha.mysite.com:8443/api/services/conversation/process" -v --http1.1 --tlsv1.2
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying ipAddress:8443...
* Connected to ha.mysite.com (ipAddress) port 8443 (#0)
* ALPN: offers http/1.1
*  CAfile: /usr/local/share/certs/ca-root-nss.crt
*  CApath: none
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN: server accepted http/1.1
* Server certificate:
*  subject: C=US; ST=California; L=San Francisco; O=Cloudflare, Inc.; CN=sni.cloudflaressl.com
*  start date: Jan  1 00:00:00 2023 GMT
*  expire date: Jan  1 23:59:59 2024 GMT
*  subjectAltName: host "ha.mysite.com" matched cert's "*.mysite.com"
*  issuer: C=US; O=Cloudflare, Inc.; CN=Cloudflare Inc ECC CA-3
*  SSL certificate verify ok.
> POST /api/services/conversation/process HTTP/1.1
> Host: ha.mysite.com:8443
> User-Agent: curl/7.85.0
> Accept: */*
> Authorization: Bearer 11111111111111111111111111
> Content-Type: application/json
> Content-Length: 50
>
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* old SSL session ID is stale, removing
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Date: Wed, 24 May 2023 01:40:38 GMT
< Content-Type: application/json
< Transfer-Encoding: chunked
< Connection: keep-alive
< CF-Cache-Status: DYNAMIC
< Report-To: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v3?s=R%2FarRvQnq2fCATFLiPnnidDSDqU76jUZEVddw5%2BP1EBBK6r7u9CzteuShp5SzbiquM2dcv4onjL29ccTqxb11T8WchTEeWSVnnoox3GdSQyS%2FedatwbaDur2Y1t9afVB76Y2whNyMA%3D%3D"}],"group":"cf-nel","max_age":604800}
< NEL: {"success_fraction":0,"report_to":"cf-nel","max_age":604800}
< Server: cloudflare
< CF-RAY: 7cc1d4e94e2f1cf0-ORD
< alt-svc: h3=":8443"; ma=86400, h3-29=":8443"; ma=86400
<
* Connection #0 to host ha.mysite.com left intact
[]

**Reply by lundyfpv** [reply]
I kinda feel like you wanted me to remove the Authorization part from the curl command so here that is as well. This gives the same error I am getting from Willow in the Home Assistant logs but that is to be expected.

curl -X POST -H "Content-Ty application/json" -d '{"text":"Turn on kitchen lights.","language":"en"}' "https://ha.mysite.com:8443/api/services/conversation/process" -v --http1.1 --tlsv1.2
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying ipAddress:8443...
* Connected to ha.mysite.com (ipAddress) port 8443 (#0)
* ALPN: offers http/1.1
*  CAfile: /usr/local/share/certs/ca-root-nss.crt
*  CApath: none
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN: server accepted http/1.1
* Server certificate:
*  subject: C=US; ST=California; L=San Francisco; O=Cloudflare, Inc.; CN=sni.cloudflaressl.com
*  start date: Jan  1 00:00:00 2023 GMT
*  expire date: Jan  1 23:59:59 2024 GMT
*  subjectAltName: host "ha.mysite.com" matched cert's "*.mysite.com"
*  issuer: C=US; O=Cloudflare, Inc.; CN=Cloudflare Inc ECC CA-3
*  SSL certificate verify ok.
> POST /api/services/conversation/process HTTP/1.1
> Host: ha.mysite.com:8443
> User-Agent: curl/7.85.0
> Accept: */*
> Content-Length: 50
> Content-Type: application/x-www-form-urlencoded
>
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* old SSL session ID is stale, removing
* Mark bundle as not supporting multiuse
< HTTP/1.1 401 Unauthorized
< Date: Wed, 24 May 2023 01:55:09 GMT
< Content-Type: text/plain; charset=utf-8
< Content-Length: 17
< Connection: keep-alive
< CF-Cache-Status: DYNAMIC
< Report-To: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v3?s=34n64ltXTfyPq%2FpkXtcX71O3bhboPn5DwpYABBxlVQTlact0P8wzKgAmjGRqQLrC8RndxV%2FeMw8%2B%2BQAsXCBG%2BThJ%2BXMT5PsqQEar1jsM4IL3w7pzr0hpNurWlyfx9U2uHXRqp7AMkQ%3D%3D"}],"group":"cf-nel","max_age":604800}
< NEL: {"success_fraction":0,"report_to":"cf-nel","max_age":604800}
< Server: cloudflare
< CF-RAY: 7cc1ea30987a44f1-ATL
< alt-svc: h3=":8443"; ma=86400, h3-29=":8443"; ma=86400
<
* Connection #0 to host ha.mysite.com left intact
401: Unauthorized

**Reply by kristiankielhofner** [reply]
What I suspect (more of a desperate hunch) may be happening here is the Willow HTTP client (from ESP-IDF) currently only supports up to TLS 1.2 (TLS 1.3 is in the next release, I believe). Cloudflare allows you to configure TLS 1.3 as the minimum version for the domain:

https://developers.cloudflare.com/ssl/edge-certificates/additional-options/minimum-tls/

Can you check this setting and try setting it to TLS 1.2 and see if that fixes this issue? While you're poking around in the dashboard, make sure you have support for Websockets enabled:

CF Dashboard -> Your domain -> Network -> Websockets

**Reply by lundyfpv** [reply]
Just want to say thanks again for your help on this, you have gone above and beyond what I expected. 

I adjusted my TLS level to 1.2 and disabled 1.3 support and verified websockets are enabled. Still not working sadly, but I did override my local DNS on my router to redirect my URL directly to my Home Assistant IP and that allowed Willow to work, sadly I couldn't access the web interface after doing that so I will have to come up with another solution. Does look like something over the Cloudflare Tunnel is causing issues. I think this is a somewhat popular method of doing remote access to Home Assistant so I will keep digging on my side an report back any solution I am able to find.

**Reply by kristiankielhofner** [reply]
Thanks!

I suppose what I'm trying to understand most about the use of Cloudflare tunnels with Willow is - why?

Home Assistant's best strength is being locally hosted. Willow is presumably in your home, local, on the same or an immediately adjacent network (VLAN or something). What's the value of communication between Willow and HA going over a CF tunnel? I can understand for access to the HA system when not at home/on the same network, that makes sense. I'm just really not understanding how/where it comes in for Willow-HA communication.

Yes, please do!

**Reply by lundyfpv** [reply, marked as answer]
You are correct Willow is in my home. It looks like the way I originally configured my Cloudflare tunnel made it where the only way to communicate with it was through the Cloudflare tunnel. I had no local access. I have reconfigured it using the new method below and everything works as expected! Thanks again for battling with me through this weird issue. Got a Tesla P4 on the way to try out WIS!

The 2 methods are below for anyone running into issues. I have included the old method if you need to reverse your steps to get back to square 1. 

[New Method](https://everythingsmarthome.co.uk/free-remote-access-for-home-assistant-with-cloudlfare/)

[Old Method](https://hodgkins.io/securing-home-assitant-with-cloudflare#allow-proxy-requests-to-home-assistant)

**Reply by kristiankielhofner** [reply]
Glad that everything worked out, let us know how things go with WIS!

**Comment by hamishcunningham** [comment]
I've had good experience with http://pagekite.net/ in the past, might be worth a look.
